### DISCUSSION TITLE
Set cursor position independently from current view position

### DISCUSSION THREAD

**Discussion by Serek16** [katspaugh/wavesurfer.js#3361] [Q&A] [answered]
Can I somehow set cursor position independently to slider position? For example when I have long waveform I would like focus on the part at e.g 0.7 of the whole waveform. And next I would like to put cursor at 0.65 without moving the view.

To better understand run this demo. You can see that when you click at the region and zoom in, it become the center of the view. But so is cursor position. In this case I want to put cursor position at the start of the region without moving region's position on the screen.

``` JS

import WaveSurfer from 'https://unpkg.com/wavesurfer.js@7/dist/wavesurfer.esm.js'
import RegionsPlugin from 'https://unpkg.com/wavesurfer.js@7/dist/plugins/regions.esm.js'

// Create own media element
const audio = new Audio()
audio.src = '/examples/audio/audio.wav'
audio.controls = true
audio.style.width = "100%"

const ws = WaveSurfer.create({
  container: '#waveform',
  waveColor: 'rgb(200, 0, 200)',
  progressColor: 'rgb(100, 0, 100)',
  media: audio,
  minPxPerSec: 1,
})

// Initialize the Regions plugin
const wsRegions = ws.registerPlugin(RegionsPlugin.create())

ws.once('decode', () => {
  wsRegions.addRegion({
    start: 9,
    end: 10,
    content: 'Click me',
    color: 'rgba(10,10,10,0.5)',
    minLength: 1,
    maxLength: 10,
    drag: false,
    resize: false,
  })
})

ws.on('ready', () => {
  // Add Audio controls
  const audioElement = document.getElementById("audioControls")
  audioElement.appendChild(audio)
})

wsRegions.on('region-clicked', (region, e) => {
  e.preventDefault()
  e.stopPropagation()
  zoomToRegion(region)
})

/*
<html>
  <div id="waveform"></div>
  <div id="audioControls"></div>
</html>
*/

function zoomToRegion(region) {
  let regionDuration = region.end - region.start

  // Zoom that the region takes up 50% of the curren waveform width
  let zoomVal = (ws.renderer.container.clientWidth * 0.5) / regionDuration
  ws.zoom(zoomVal)

  ws.setTime(region.start + regionDuration / 2)

  refreshWaveform()
}
```

**Comment by katspaugh** [comment, marked as answer]
Not sure I fully understand what you want to achieve, but you could try setting `autoScroll: false`.

**Reply by Serek16** [reply]
I was about to write that I tried this already and it didn't work for but I tinkered with some buttons and then I came to conslusion that
This, my first idea, it didn't work for me:

```JS
function zoomToRegion(region) {
  let regionDuration = region.end - region.start
  [...]
  ws.setTime(region.start + regionDuration / 2)
  ws.options.autoScroll = false
  ws.setTime(region.start)
  ws.options.autoScroll = true
}
```

![Zrzut ekranu 2023-11-28 094638](https://github.com/katspaugh/wavesurfer.js/assets/50013854/ffd7a959-fec8-4057-bef3-b10994298be0)


But it does:

```JS
async function zoomToRegion(region) {
  let regionDuration = region.end - region.start
  [...]
  ws.setTime(region.start + regionDuration / 2)

  await delay(10)

  ws.options.autoScroll = false
  ws.setTime(region.start)
  ws.options.autoScroll = true
}

function delay(time) {
  return new Promise(resolve => setTimeout(resolve, time));
}
```

![Zrzut ekranu 2023-11-28 092244](https://github.com/katspaugh/wavesurfer.js/assets/50013854/b348129c-7294-4fda-9bbf-2f08349a667d)

**Reply by katspaugh** [reply]
Yes, updating the time is asynchronous, so you should set autoScroll back to true either after a delay, or on timeupdate.

**Reply by jagthedrummer** [reply]
@katspaugh would it be possible to have `setTime` return a promise so that calling code can know when it's safe to proceed? I'm trying to do exactly this same thing and the length of the delay that I need seems to be highly variable across browsers and system loads. To get it to work consistently across my development environments I'm having to use a delay of 200ms. Not a huge deal, but it's long enough to be noticeable, and in some circumstances it's _way_ longer than is really needed.

**Reply by katspaugh** [reply]
That would complicate the code somewhat, as setTime would have to create a one-time subscription to the seeked event internally and unsubscribe immediately and additionally on destroy.
I’m open to adding it though.

**Reply by jagthedrummer** [reply]
It may be overkill. I don't really ever need to be able to `await` a call to `setTime` except for in this one instance of trying to make a "focused" player that shows only a particular segment of an audio file. If there were some sort of `setViewWindow` method that would alter the display without the overhead of interfacing with the actual audio element that might be a cleaner approach.

**Reply by katspaugh** [reply]
Yeah, that’d be good. Something that sets the scroll position and zoom level from two timestamps.

**Comment by jagthedrummer** [comment]
https://github.com/katspaugh/wavesurfer.js/pull/3547 added a `setScrollTime` method that can be used to adjust the viewing window independently of the cursor. It should be available in the next release after 7.7.2 (probably either 7.7.3 or 7.8.0).
